google-cloud-storage/src/test/java/com/google/cloud/storage/GrpcStorageImplUploadRetryTest.java

```diff
@@ -0,0 +1,255 @@
+        FakeServer server = FakeServer.of(service);
+        Storage s = server.getGrpcStorageOptions().getService();
+        InputStream in = Channels.newInputStream(tmpFile.reader())) {
+      BlobInfo info = BlobInfo.newBuilder("buck", "obj").build();
+      s.createFrom(info, in, BlobWriteOption.doesNotExist());
+    }
+
+    assertThat(service.returnError.get()).isFalse();
+  }
+
+  private static final class Direct {
+    private static final Object obj =
+        Object.newBuilder().setBucket("projects/_/buckets/buck").setName("obj").build();
+    private static final WriteObjectSpec spec =
+        WriteObjectSpec.newBuilder().setResource(obj).setIfGenerationMatch(0).build();
+
+    private static final ChecksummedData checksummedData =
+        TestUtils.getChecksummedData(ByteString.copyFrom(bytes), Hasher.enabled());
+    private static final WriteObjectRequest req1 =
+        WriteObjectRequest.newBuilder()
+            .setWriteObjectSpec(spec)
+            .setChecksummedData(checksummedData)
+            .setObjectChecksums(ObjectChecksums.newBuilder().setCrc32C(checksummedData.getCrc32C()))
+            .setFinishWrite(true)
+            .build();
+    private static final WriteObjectResponse resp1 =
+        WriteObjectResponse.newBuilder()
+            .setResource(obj.toBuilder().setSize(objectContentSize))
+            .build();
+
+    private static final class FakeService extends DirectWriteService {
+      private final AtomicBoolean returnError;
+
+      private FakeService(AtomicBoolean returnError) {
+        super(
+            (obs, reqs) -> {
+              if (reqs.equals(ImmutableList.of(req1))) {
+                if (returnError.get()) {
+                  returnError.compareAndSet(true, false);
+                  obs.onError(TestUtils.apiException(Code.INTERNAL, "should retry"));
+                } else {
+                  obs.onNext(resp1);
+                  obs.onCompleted();
+                }
+              } else {
+                obs.onError(
+                    TestUtils.apiException(Code.PERMISSION_DENIED, "Unexpected request chain."));
+              }
+            });
+        this.returnError = returnError;
+      }
+
+      // a bit of constructor lifecycle hackery to appease the compiler
+      // Even though the thing past to super() is a lazy function, the closing over of the outer
+      // fields happens earlier than they are available. To side step this fact, we provide the
+      // AtomicBoolean as a constructor argument which can be closed over without issue, and then
+      // bind it to the class field after super().
+      static Direct.FakeService create() {
+        return new Direct.FakeService(new AtomicBoolean(true));
+      }
+    }
+  }
+
+  private static final class Resumable {
+
+    private static final String uploadId = "upload-id";
+
+    private static final Object obj =
+        Object.newBuilder().setBucket("projects/_/buckets/buck").setName("obj").build();
```

Comment:
Looks like `"projects/_/buckets/buck"` values can be pulled up into a single variable used by the test file instead of inline literals. You may have already considered this and I would like to understand.

Reply:
hehe, nope copy pasted. I'll create a common variable